ui/raidboss/emulator/overrides/RaidEmulatorWatchCombatantsOverride.ts

```diff
@@ -0,0 +1,89 @@
+import Util, { WatchCombatantParams } from '../../../../resources/util';
+import { OverlayHandlerRequests, OverlayHandlerResponseTypes } from '../../../../types/event';
+import AnalyzedEncounter from '../data/AnalyzedEncounter';
+import CombatantTracker from '../data/CombatantTracker';
+import RaidEmulator from '../data/RaidEmulator';
+
+import RaidEmulatorOverlayApiHook from './RaidEmulatorOverlayApiHook';
+
+type Watch = {
+  lastCheck: number;
+  params: WatchCombatantParams;
+  msg: OverlayHandlerRequests['getCombatants'];
+  cancel: boolean;
+  start: number;
+  promise: Promise<void>;
+  res: () => void;
+  rej: () => void;
+  func: (ret: OverlayHandlerResponseTypes['getCombatants']) => boolean;
+};
+
+export default class RaidEmulatorWatchCombatantsOverride {
+  private watches: Watch[] = [];
+
+  constructor(private emulator: RaidEmulator, private overlayHook: RaidEmulatorOverlayApiHook) {
+    Util.setWatchCombatantOverride((params, func) => {
+      // To avoid having to undefined-check the params declare watch as partial here and just cast
+      // to non-partial when pushing to array since all props are set here properly but can't be
+      // set in the initializer
+      const watch: Partial<Watch> = {
+        lastCheck: 0,
+        params: params,
+        cancel: false,
+        start: 0,
+        func: func,
+        msg: {
+          call: 'getCombatants',
+          ...params,
+        },
+      };
+
+      watch.promise = new Promise<void>((res, rej) => {
+        watch.res = res;
+        watch.rej = rej;
+      });
+      this.watches.push(watch as Watch);
```

Comment:
I'm not sure what you mean about undefined checks.

```suggestion
    Util.setWatchCombatantOverride((params, func) => {
      const watch: Watch = {
        lastCheck: 0,
        params: params,
        cancel: false,
        start: 0,
        func: func,
        msg: {
          call: 'getCombatants',
          ...params,
        },
        res: () => {/* empty */},
        rej: () => {/* empty */},
        promise: new Promise<void>((res, rej) => {
          watch.res = res;
          watch.rej = rej;
        }),
      };

      this.watches.push(watch);
```

Reply:
Doesn't this run into the issue of `watch` not being set when the promise attempts to set `watch.res`?

Reply:
If that's true, then `let watch: Watch; watch = { etc` should do the trick, right?

Reply:
I'd expect eslint to complain about that but I'm not able to check right now (still updating my linux install). If not, I'll correct it tomorrow.

Reply:
FWIW, typescript did not complain at me about the original, but that doesn't mean that there's not a bug there too.

Reply:
Yeah, eslint complains about that syntax.

![image](https://user-images.githubusercontent.com/6119598/126178176-9a331f64-e03f-4187-a191-bf10cc0bf34f.png)

The problem with the original suggestion is I think the same thing we ran into in #3156, but I still need to test to be sure.

Reply:
Had a chance to test the original suggestion as well, it throws the same error from the bug I linked, as I thought it would:
```
Uncaught (in promise) ReferenceError: can't access lexical declaration 'watch' before initialization
```

Reply:
I think it is better to disable eslint than to assert something (possibly false in the future) to TypeScript.  That said, will https://eslint.org/docs/rules/prefer-const#ignorereadbeforeassign fix that error?

Reply:
No, unfortunately, setting that won't fix the eslint error in the case of `let watch; watch = ...`.

And the original suggestion's error is flat out a browser js error, not an eslint/ts error.

Reply:
Oh, you're right.  I misread that eslint.

That said, I still think it is better to disable eslint's prefer-const for that line than to add a type assertion.

Reply:
Trying to clean up some of these old PRs. I still need to test the latest changes, but I think this is what you were suggesting?

Reply:
Hmm, after testing, even with these changes it's still throwing an exception.
![image](https://user-images.githubusercontent.com/6119598/164313128-e5e111b0-e73e-49f6-9bea-823d812aea47.png)

Reply:
Attempt number 3. This one works in testing and doesn't need the eslint disable, but it does require more indenting.

Reply:
Mmm, I was trying to say that you could use `let` here if you needed to for correctness, but just disable the eslint about using const, vs using a `watch as Watch` type assertion.  I can't see what the old code was to leave a snippet, but that's what I meant.

As a general, rule, I think narrow eslint disabling > broad eslint disabling > type assertions as preferred mechanisms for dealing with issues.  Type assertions just feel like working around TypeScript and can hide issues in the future during refactoring.  That's really what I was trying to get at.

Reply:
Yeah, I tried using `let` but it still errored at runtime, I'm guessing due to execution priority (it was executing the promise's anonymous function parameter before actually assigning the generated object to the `watch` variable).

Reply:
For reference, here's the previous iteration where `let` was still erroring:
https://github.com/quisquous/cactbot/blob/254afec6c6271f6132dda33653743cd12e65c5d4/ui/raidboss/emulator/overrides/RaidEmulatorWatchCombatantsOverride.ts#L23-L51

Reply:
> Yeah, I tried using let but it still errored at runtime

Huh!!

Thanks for the link.